### PR TITLE
Upgrade to support electron 4 (2)

### PR DESCRIPTION
**Pull request by alexstrat**
Finish #354 by implementing the workaround of https://github.com/electron/spectron/issues/346#issuecomment-483607219

### CODE REVIEW

File: test/application-test.js [hunk 1]
```diff
@@ -137,7 +137,7 @@ describe('application loading', function () {
     it('gets the render process console logs and clears them', function () {
       return app.client.waitUntilWindowLoaded()
         .getRenderProcessLogs().then(function (logs) {
-          expect(logs.length).to.equal(3)
+          expect(logs.length).to.equal(4)
```
Comment (alexstrat) [inline test/application-test.js]: I have not checked, but probably an Electron warning is fired.

File: test/multi-window-test.js [hunk 2]
```diff
@@ -24,15 +24,15 @@ describe('multiple windows', function () {
   it('launches the application', function () {
     return app.client
       .getWindowCount().should.eventually.equal(2)
-      .windowByIndex(1)
+      .windowByIndex(0)
```
Comment (alexstrat) [inline test/multi-window-test.js]: Not sure why the order has changed, but don't think this change breaks the idea of the test.

Comment (alexstrat) [reply test/multi-window-test.js]: Actually, it looks like we can't rely on the order of `windowhandles`

https://github.com/webdriverio/webdriverio/blob/567f8778bc8a66a2a68ad9e12da2c81e86fabe47/packages/webdriver/protocol/webdriver.json#L192

So I rewrote the test to be independent of the order of the returned windows.

**Comment by alexstrat** [comment]
Failing on windows:

```txt
 1) window commands
       webContents.selectAll()
         selects all the text on the page:
     AssertionError: expected '' to include 'Hello'
```

Have no clue so far: is `getSelectedText` or `selectAll` not working?

File: lib/application.js [hunk 3]
```diff
@@ -149,6 +149,11 @@ Application.prototype.createClient = function () {
       args.push('spectron-env-' + name + '=' + self.env[name])
     }
 
+    // Port to communicate on. Required starting with ChromeDriver v2.41.
+    // https://bitbucket.org/chromiumembedded/cef/issues/2476/cef-webdriver-instructions-will-need-to
+    const randomPort = Math.floor(Math.random() * (9999 - 9000) + 9000)
```
Comment (MarshallOfSound) [inline lib/application.js]: This is not the right port to use, we should be using port `0` and reading the `DevToolsActivePort` file to get the chosen port 🤔

Comment (alexstrat) [reply lib/application.js]: From the link above:

> It also doesn't work with CEF. The reason is that CEF's devtools implementation refuses to handle the remote-debugging-port=0 use case. See this code: https://bitbucket.org/chromiumembedded/cef/src/8e9d736cda2015b706e183bfd8f7fd63a2cd5396/libcef/browser/devtools_manager_delegate.cc#lines-75

So I considered that, in Electron (CEF right?), `remote-debugging-port=0` won't work and have not given it a try.

I can if you think I should!

Comment (MarshallOfSound) [reply lib/application.js]: @alexstrat I fixed `--remote-debugging-port=0` a while back in Electron -> https://github.com/electron/electron/pull/17800

Comment (alexstrat) [reply lib/application.js]: Oh yes! I missed that indeed!

However, it looks like the fix is only available for electron >=5 and this PR is for electron 4.

Should I either:
- backport the fix in Electron 4
- keep this random port solution until Spectron 7 (Electron 5) and add a comment here about it
- try the `remote-debugging-port=0` here and see if it works (is there a reason for it to work?)

Comment (alexstrat) [reply lib/application.js]: I started looking into electron 5 and I'm not sure we need to read `DevToolsActivePort` actually. Indeed currently `webdriver@4` does not use the [`devtools-service`](https://github.com/webdriverio-boneyard/wdio-devtools-service) (no trace in dependencies, etc..) which is the only component that'll need access to devtools.

I was able to get it running just by removing `args.push('remote-debugging-port=${randomPort}')`.

Comment (alexstrat) [reply lib/application.js]: I chose to keep the random port solution until Spectron 7. #364

**Comment by alexstrat** [comment]
Ready for 2nd review! 👐

**Comment by jonbeller** [comment]
I am using spectron@5.0.0 with electron@4.0.4. What might be broken for me? Sorry if this is not the right place to ask!

File: lib/application.js [hunk 4]
```diff
@@ -153,6 +153,20 @@ Application.prototype.createClient = function () {
       args.push(arg)
     })
 
+    // Since ChromeDriver v2.41, ChromeDriver with CEF will only connect
```
Comment (codebytere) [inline lib/application.js]: Electron doesn't use CEF :) Do you mean this is broken here in a similar way it's broken in CEF?

**Comment by codebytere** [comment]
Looks like
```
    webContents.selectAll()
      1) selects all the text on the page
```
is still failing 🤔

**Comment by alexstrat** [comment]
@codebytere yes, indeed, but it was failing since electron 3 #334 see 25d492dc02e6e29e03f09b7613390fedba4d9068's [Appeyor build](https://ci.appveyor.com/project/electron-bot/spectron/builds/19697638/job/jkvru287045h228x).

I've done some research about it but I could not find a clue and I can hardly test in Windows. If you think it's necessary for merging this PR, I can have a a look.

File: lib/application.js (hunk 4 above)
Comment (alexstrat) [reply lib/application.js]: Ah yes, it was not precise, but yes, I guess it's similar behavior.

**Comment by alexstrat** [comment]
If ok, I think we can merge (i can't) and release?

**Pull request merged**